packages/gatsby-source-contentful/src/gatsby-node.js

```diff
@@ -49,9 +49,18 @@ exports.sourceNodes = async (
     process.env.GATSBY_CONTENTFUL_OFFLINE === `true` &&
     process.env.NODE_ENV !== `production`
   ) {
-    getNodes()
-      .filter(n => n.internal.owner === `gatsby-source-contentful`)
-      .forEach(n => touchNode({ nodeId: n.id }))
+    let contentfulNodes = getNodes().filter(
+      n => n.internal.owner === `gatsby-source-contentful`
+    )
+    contentfulNodes.forEach(n =>
+      touchNode({
+        nodeId: n.id,
+      })
+    )
+    contentfulNodes
+      .map(n => n.localFile___NODE)
+      .filter(id => id)
+      .forEach(id => touchNode({ nodeId: id }))
```

Comment:
Consider what happens at scale; a site with 100k+ pages will have more than 100k nodes. These loops are expensive and add up. Instead why not take a step back and consider whether it's still worth to keep the array lambda methods.

How about something like this instead?

```suggestion
getNodes().forEach(node => {
  if (node.internal.owner !== `gatsby-source-contentful`) {
    return;
  }

  touchNode({nodeId: node.id});

  if (node.localFile___NODE) {
    // Prevent GraphQL type inference from crashing on this property
    touchNode({nodeId: node.localFile___NODE});
  }
});
```

Single loop and array instead of five loops and 4 arrays (start, filter, map, filter).

Reply:
Sounds good; applied this and verified still fixes the issue locally.